WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift

```diff
@@ -112,37 +131,96 @@ private extension SettingsViewController {
-            onCompletion()
+        sitesResultsController.onDidResetContent = { [weak self] in
+            guard let self = self else {
+                return
+            }
+
+            self.refetchAllResultsControllers()
+            onReload()
+        }
+
+        try? sitesResultsController.performFetch()
+
+        guard paymentGatewayAccountsResultsController != nil else {
             return
         }
 
-        let action = WCPayAction.loadAccount(siteID: siteID) { [weak self] result in
+        paymentGatewayAccountsResultsController?.onDidChangeContent = {
+            onReload()
+        }
+
+        paymentGatewayAccountsResultsController?.onDidResetContent = { [weak self] in
             guard let self = self else {
                 return
             }
+            self.refetchAllResultsControllers()
+            onReload()
+        }
 
-            switch result {
-            case .failure:
-                self.canCollectPayments = false
-            case .success(let account):
-                self.canCollectPayments = ServiceLocator.featureFlagService.isFeatureFlagEnabled(.cardPresentPayments) &&
-                    account.isCardPresentEligible
-            }
+        try? paymentGatewayAccountsResultsController?.performFetch()
+    }
 
-            onCompletion()
+    /// Refetching all the results controllers is necessary after a storage reset in `onDidResetContent` callback and before reloading UI that
+    /// involves more than one results controller.
+    ///
+    private func refetchAllResultsControllers() {
+        try? sitesResultsController.performFetch()
+        guard paymentGatewayAccountsResultsController != nil else {
+            return
+        }
+        try? paymentGatewayAccountsResultsController?.performFetch()
+    }
+
+    /// Ask the PaymentGatewayAccountStore to loadAccounts from the network and update storage
+    ///
+    private func loadPaymentGatewayAccounts() {
+        guard let siteID = self.siteID else {
+            return
         }
 
+        /// No need for a completion here. We will be notified of storage changes in `onDidChangeContent`
+        ///
+        let action = PaymentGatewayAccountAction.loadAccounts(siteID: siteID) {_ in
+        }
         ServiceLocator.stores.dispatch(action)
     }
 
+    /// Update our list of sites from the sitesResultsController
+    ///
+    private func updateSites() {
+        sites = sitesResultsController.fetchedObjects
+    }
+
+    /// Determine if any payment gateway account for this site supports card present payments
+    ///
+    private func checkAvailabilityForPayments() {
+        guard ServiceLocator.featureFlagService.isFeatureFlagEnabled(.cardPresentPayments) else {
+            canCollectPayments = false
+            return
+        }
+
+        paymentGatewayAccounts = paymentGatewayAccountsResultsController?.fetchedObjects ?? []
+
+        guard paymentGatewayAccounts.isNotEmpty else {
```

Comment:
This might not be necessary. If we just set `cancollectPayments = paymentGatewayAccounts.contains...` this method might be a bit more compact

Reply:
It could. I thought of that too. I kinda wanted to make it explicit that an empty array is expected, but I probably don't need to be that explicit. Will update.

Reply:
Fixed in https://github.com/woocommerce/woocommerce-ios/pull/4398/commits/bef78fec49d56896f723312c544a5f408b28b311